src/Tuura/Concept/FSM/Translation.hs

```diff
@@ -209,9 +232,22 @@ fsmarcxToFsmarc arc = FsmArc newSourceEnc (transx arc) newDestEnc
     where newSourceEnc = (encToInt . srcEncx) arc
           newDestEnc = (encToInt . destEncx) arc
 
-removeInvariants :: [FsmArcX a] -> [[Tristate]] -> [FsmArcX a]
-removeInvariants xs inv = filter (\x -> not (destEncx x `elem` inv || srcEncx x `elem` inv)) xs
+removeUnreachables :: [FsmArc a] -> [Int] -> [FsmArc a]
+removeUnreachables xs reachables = filter (\s -> (destEnc s `elem` reachables && srcEnc s `elem` reachables)) xs
 
 -- Produce all arcs with all X's resolved
-createAllArcs :: Ord a => [([Transition a], Transition a)] -> [[Tristate]] -> [FsmArc a]
-createAllArcs xs = map fsmarcxToFsmarc . removeInvariants (expandAllXs (createArcs xs))
+createAllArcs :: Ord a => [([Transition a], Transition a)] -> [FsmArc a]
+createAllArcs = map fsmarcxToFsmarc . expandAllXs . createArcs
+
+findReachables :: [FsmArc a] -> Int -> [Int]
+findReachables allArcs initialState = nubOrd (visit initialState allArcs [])
+
+visit :: Int -> [FsmArc a] -> [Int] -> [Int]
+visit state allArcs visited = [state] ++
+        if (nonVisited /= [])
+        then concatMap (\s -> visit s allArcs (visited ++ [state] ++ nonVisited)) nonVisited
+        else []
+    where
+      indices = elemIndices state (map srcEnc allArcs)
+      destArcs = map ((map destEnc allArcs) !!) indices
```

Comment:
`!!` is very slow. Perhaps, you could implement this using a `Data.Set` or `Data.Map`?